cpp/include/cudf/utilities/error.hpp

```diff
@@ -104,15 +104,20 @@ inline void throw_cuda_error(cudaError_t error, const char* file,
                   cudaGetErrorName(error) + " " + cudaGetErrorString(error)});
 }
 
-inline void check_stream(cudaStream_t stream, const char* file,
-                         unsigned int line) {
+template <bool sync_on_stream>
+inline void check_cuda(char const* file, unsigned int line,
+                       cudaStream_t stream=0) {
   cudaError_t error{cudaSuccess};
-  error = cudaStreamSynchronize(stream);
-  if (cudaSuccess != error) {
-    throw_cuda_error(error, file, line);
+  if (sync_on_stream) {
+    error = cudaStreamSynchronize(stream);
+    if (cudaSuccess == error) {
+      error = cudaGetLastError();
```

Comment:
Why call `cudaGetLastError()` again after `cudaStreamSynchronize()`? If there were any latent CUDA errors, I believe they would have been picked up in the `cudaStreamSynchronize()`.

Reply:
In fact, I don't know why I didn't think of this earlier, but this function should be simplified to just do:
```c++
if(sync_on_stream)
   CUDA_TRY(cudaStreamSynchronize(stream));
else
   CUDA_TRY(cudaGetLastError);
```

Reply:
> Why call `cudaGetLastError()` again after `cudaStreamSynchronize()`? If there were any latent CUDA errors, I believe they would have been picked up in the `cudaStreamSynchronize()`.

I didn't know that, and cudaStreamSynchronize also clears the error? If not, shouldn't we still need to call cudaGetLastError()?

Reply:
And even if cudaStreamSynchronize currently returns any latent CUDA errors, if it's not publicly documented behavior of cudaStreamSynchronize, isn't it dangerous to rely on such behavior?

https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__STREAM.html#group__CUDART__STREAM_1g82b5784f674c17c6df64affe618bf45e

```
Note that this function may also return error codes from previous, asynchronous launches.
```

The only thing I can find is the above, and this does not say this function is guaranteed to return error codes from previous, asynchronous launches.

Reply:
I think that we want to call cudaGetLastError as well as cudaStreamSynchronize() because the former resets the error state to cudaSuccess. We don't want to leave the context in an error state, right?

Reply:
Also, the classic answer: https://stackoverflow.com/a/14038590/749748

Reply:
My reading is that `cudaErrorSyncDepthExceeded` can only happen in dynamic parallel launches, which we don't use currently.

In general, just one call to cudaGetLastError() and cudaStreamSynchronize() is sufficient.

Reply:
If I were to implement this function, I would have done it in this order:

```
template <bool sync_on_stream>
 inline void check_cuda(char const* file, unsigned int line,
                        cudaStream_t stream=0) {
  cudaError_t error = cudaGetLastError(); // Check for any existing error
  if (sync_on_stream && (cudaSuccess == error)) {
    error = cudaStreamSynchronize(stream); // Wait for work in stream to finish and check for new errors
  }
  if (cudaSuccess != error) {
    throw_cuda_error(error, file, line);
  }
}
```

The `cudaGetLastError()` will get any immediately posted errors, such as launch configuration errors. The sync will catch errors in kernels, e.g. accessing invalid address.

Reply:
> If I were to implement this function, I would have done it in this order:
> 
> ```
> template <bool sync_on_stream>
> inline void check_cuda(char const* file, unsigned int line,
>                         cudaStream_t stream=0) {
>   cudaError_t error = cudaGetLastError(); // Check for any existing error
>   if (sync_on_stream && (error == cudaSuccess)) {
>     error = cudaStreamSynchronize(stream); // Wait for work in stream to finish and check for new errors
>   }
>   if (error != cudaSuccess) {
>     throw_cuda_error(error, file, line);
>   }
> }
> ```
> 
> The `cudaGetLastError()` will get any immediately posted errors, such as launch configuration errors. The sync will catch errors in kernels, e.g. accessing invalid address.

What about clearing the last error? If an error occurs between cudaGetLastError() and cudaStreamSynchronize(), it will remain uncleared in the above code.

Reply:
> If an error occurs between cudaGetLastError() and cudaStreamSynchronize(), it will remain uncleared in the above code.

Won't `cudaStreamSynchronize()` clear it?

Reply:
Did y'all not like the suggestion above?
```c++
#ifndefine NDEBUG 
#define CHECK_CUDA(stream)   CUDA_TRY(cudaStreamSynchronize(stream));
#else 
#define CHECK_CUDA(stream)   CUDA_TRY(cudaGetLastError());
#endif 
```
After https://github.com/rapidsai/cudf/issues/3510 is addressed, then `CUDA_TRY` will take care of calling `cudaGetLastError()` to clear any pending errors.

Reply:
https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__STREAM.html#group__CUDART__STREAM_1g82b5784f674c17c6df64affe618bf45e

The documentation above about cudaStreamSynchronize does not say anything about clearing error,

while the documentation below about cudaGetLastError clearly states that

https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__ERROR.html#group__CUDART__ERROR_1g3529f94cb530a83a76613616782bd233

"Returns the last error that has been produced by any of the runtime calls in the same host thread and **resets it to cudaSuccess**."

Is there any other document about cudaStreamSynchronize I am missing (or is this an undocumented behavior)?

>

Reply:
Ahh, yes Jake's solution is good if CUDA_TRY is changed to getLastError().

Reply:
@seunghwak do you want to just knock out #3510 in this same PR?

Reply:
> Did y'all not like the suggestion above?
> 
> ```c++
> #ifndefine NDEBUG 
> #define CHECK_CUDA(stream)   CUDA_TRY(cudaStreamSynchronize(stream));
> #else 
> #define CHECK_CUDA(stream)   CUDA_TRY(cudaGetLastError());
> #endif 
> ```
> 
> After #3510 is addressed, then `CUDA_TRY` will take care of calling `cudaGetLastError()` to clear any pending errors.

If this function is not expected to clear error,

Then, we may go with the code below (if cudaStreamSynchronize is guaranteed to return error from previous calls, the documentation only says it **may** return error from previous calls).

```c++
#ifndefine NDEBUG 
#define CHECK_CUDA(stream)   CUDA_TRY(cudaStreamSynchronize(stream));
#else 
#define CHECK_CUDA(stream)   CUDA_TRY(cudaPeekLastError());
#endif 
```

But if we expect the error to be cleared after properly handling exceptions from CHECK_CUDA_ERROR, I think we need to call cudaGetLastError() at the end.

Reply:
The purpose of #3510 is to add `cudaGetLastError()` to `CUDA_TRY` to reset the error code to `cudaSuccess`. I believe that addresses your concern.

Reply:
> @seunghwak do you want to just knock out #3510 in this same PR?

Yes, I can do that. So, you guys expect this function to not clear error? If that's the case,

we may call cudaPeekLastError() instead of cudaGetLastError() and we may also rename this to PEEK_CUDA to make it clear. This function clearing error on release build and not clearing in debug build sounds a bit confusing to me.

Reply:
I think it should clear the error.

Reply:
If `CHECK_CUDA` is defined as 
```
#ifndefine NDEBUG 
#define CHECK_CUDA(stream)   CUDA_TRY(cudaStreamSynchronize(stream));
#else 
#define CHECK_CUDA(stream)   CUDA_TRY(cudaPeekLastError());
#endif 
```

and #3510 is completed, then `CHECK_CUDA` will clear errors in both release and debug builds because `CUDA_TRY` will always clear errors.

Reply:
Got it, and this makes sense to me as long as we can assume cudaStreamSyncrhonize is guaranteed to return any previous errors (the documentation only says it **may**, but I assume this function will not return cudaSuccess as long as there is any pending errors.).

Reply:
Well, we can't really do any more than this. :)